atoma-proxy/src/server/middleware.rs

```diff
@@ -1882,6 +1901,72 @@ pub mod auth {
         // NOTE: When the `lock_guard` goes out of scope, it ensures that the `DashMap` entry is removed,
         // even if the `acquire_new_stack` returned an error, previously, as this is handled at drop time.
     }
+
+    /// Gets a stack from the state manager for a given model and user ID.
+    ///
+    /// This function sends a request to the state manager to retrieve a stack that can handle the
+    /// given model and user ID. It returns the stack if found, otherwise it returns None.
+    ///
+    /// # Arguments
+    /// * `state` - The state of the proxy
+    /// * `model` - The name/identifier of the AI model being requested
+    /// * `user_id` - The ID of the user making the request
+    /// * `free_compute_units` - The number of free compute units (tokens) needed for the request
+    /// * `is_confidential` - Whether the request is confidential
+    /// * `endpoint` - The API endpoint being accessed
+    ///
+    /// # Returns
+    /// * `Result<Option<SelectedNodeMetadata>>` - The stack if found, otherwise None   
+    ///
+    /// # Errors
+    /// * `AtomaProxyError::InternalError` - Failed to send or receive message to the state manager
+    #[instrument(
+        level = "info",
+        skip_all,
+        fields(model =%model, user_id =%user_id, free_compute_units =%free_compute_units, is_confidential =%is_confidential, endpoint =%endpoint),
+        err
+    )]
+    pub async fn get_node_metadata_from_state_manager(
+        state: &ProxyState,
+        model: &str,
+        user_id: i64,
+        free_compute_units: i64,
+        is_confidential: bool,
+        endpoint: &str,
+    ) -> Result<Option<SelectedNodeMetadata>> {
+        let (result_sender, result_receiver) = oneshot::channel();
+        state
+            .state_manager_sender
+            .send(AtomaAtomaStateManagerEvent::GetStacksForModel {
+                model: model.to_string(),
+                user_id,
+                free_compute_units,
+                is_confidential,
+                result_sender,
+            })
+            .map_err(|err| AtomaProxyError::InternalError {
+                message: format!("Failed to send GetStacksForModel event: {err:?}"),
+                client_message: None,
+                endpoint: endpoint.to_string(),
+            })?;
+        let maybe_stack = result_receiver
+            .await
+            .map_err(|err| AtomaProxyError::InternalError {
```

Comment:
Why do we need this duplicate?

Reply:
Because one result is about channel closed, the other result consists of possible db failures.